internal/venice-client-common/src/main/java/com/linkedin/venice/stats/metrics/MetricEntityStateTwoEnums.java

```diff
@@ -0,0 +1,179 @@
+    validateRequiredDimensions(metricEntity, null, baseDimensionsMap, enumTypeClass1, enumTypeClass2);
+    this.enumTypeClass1 = enumTypeClass1;
+    this.enumTypeClass2 = enumTypeClass2;
+    this.attributesEnumMap = createAttributesEnumMap(metricEntity, otelRepository, baseDimensionsMap);
+  }
+
+  /** Factory method with named parameters to ensure the passed in enumTypeClass are in the same order as E */
+  public static <E1 extends Enum<E1> & VeniceDimensionInterface, E2 extends Enum<E2> & VeniceDimensionInterface> MetricEntityStateTwoEnums<E1, E2> create(
+      MetricEntity metricEntity,
+      VeniceOpenTelemetryMetricsRepository otelRepository,
+      Map<VeniceMetricsDimensions, String> baseDimensionsMap,
+      Class<E1> enumTypeClass1,
+      Class<E2> enumTypeClass2) {
+    return new MetricEntityStateTwoEnums<>(
+        metricEntity,
+        otelRepository,
+        baseDimensionsMap,
+        enumTypeClass1,
+        enumTypeClass2);
+  }
+
+  /** Overloaded Factory method for constructor with Tehuti parameters */
+  public static <E1 extends Enum<E1> & VeniceDimensionInterface, E2 extends Enum<E2> & VeniceDimensionInterface> MetricEntityStateTwoEnums<E1, E2> create(
+      MetricEntity metricEntity,
+      VeniceOpenTelemetryMetricsRepository otelRepository,
+      TehutiSensorRegistrationFunction registerTehutiSensorFn,
+      TehutiMetricNameEnum tehutiMetricNameEnum,
+      List<MeasurableStat> tehutiMetricStats,
+      Map<VeniceMetricsDimensions, String> baseDimensionsMap,
+      Class<E1> enumTypeClass1,
+      Class<E2> enumTypeClass2) {
+    return new MetricEntityStateTwoEnums<>(
+        metricEntity,
+        otelRepository,
+        registerTehutiSensorFn,
+        tehutiMetricNameEnum,
+        tehutiMetricStats,
+        baseDimensionsMap,
+        enumTypeClass1,
+        enumTypeClass2);
+  }
+
+  /**
+    * Creates an EnumMap of {@link Attributes} for each possible value of the dynamic dimensions
+   * {@link #enumTypeClass1} and {@link #enumTypeClass2}
+   */
+  private EnumMap<E1, EnumMap<E2, Attributes>> createAttributesEnumMap(
+      MetricEntity metricEntity,
+      VeniceOpenTelemetryMetricsRepository otelRepository,
+      Map<VeniceMetricsDimensions, String> baseDimensionsMap) {
+    if (!emitOpenTelemetryMetrics()) {
+      return null;
+    }
+
+    EnumMap<E1, EnumMap<E2, Attributes>> attributesEnumMap = new EnumMap<>(enumTypeClass1);
+    Map<VeniceMetricsDimensions, String> additionalDimensionsMap = new HashMap<>();
+    for (E1 enumConst1: enumTypeClass1.getEnumConstants()) {
+      additionalDimensionsMap.put(enumConst1.getDimensionName(), enumConst1.getDimensionValue());
+      EnumMap<E2, Attributes> mapE2 = new EnumMap<>(enumTypeClass2);
+      attributesEnumMap.put(enumConst1, mapE2);
+      for (E2 enumConst2: enumTypeClass2.getEnumConstants()) {
+        additionalDimensionsMap.put(enumConst2.getDimensionName(), enumConst2.getDimensionValue());
```

Comment:
Is it possible that `E1` and `E2` have dimension names that clash? What would happen then? Can you write a unit test for this scenario, if we don't already have one?

Reply:
Thanks for the catch. Added some extra checks to detect this clashes and tests for the same.